src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDatabaseMetaData.java

```diff
@@ -602,77 +605,181 @@ private static String EscapeIDName(String inID) throws SQLServerException {
+                rs.getColumn(7).setFilter(new ZeroFixupFilter());
+                rs.getColumn(8).setFilter(new ZeroFixupFilter());
+                rs.getColumn(16).setFilter(new ZeroFixupFilter());
+            } catch (SQLException e) {
+                if (null != pstmt) {
+                    try {
+                        pstmt.close();
+                    } catch (SQLServerException ignore) {
+                        if (loggerExternal.isLoggable(Level.FINER) && Util.isActivityTraceOn()) {
+                            loggerExternal.finer("findColumn() threw an exception when attempting to close PreparedStatement");
+                        }
+                    }
+                }
+                throw e;
+            } finally {
+                if (null != originalCatalog) {
+                    connection.setCatalog(originalCatalog);
+                }
+            }
 
-        // sp_columns supports wild carding schema table and columns
-        String column = EscapeIDName(col);
-        table = EscapeIDName(table);
-        schema = EscapeIDName(schema);
-
-        /*
-         * sp_columns [ @table_name = ] object [ , [ @table_owner = ] owner ] [ , [ @table_qualifier = ] qualifier ] [ ,
-         * [ @column_name = ] column ] [ , [ @ODBCVer = ] ODBCVer ]
-         */
-        String[] arguments;
-        if (connection.isKatmaiOrLater())
-            arguments = new String[6];
-        else
-            arguments = new String[5];
-        arguments[0] = table;
-        arguments[1] = schema;
-        arguments[2] = catalog;
-        arguments[3] = column;
-        if (connection.isKatmaiOrLater()) {
-            arguments[4] = "2"; // give information about everything including
-                                // sparse columns
-            arguments[5] = "3"; // odbc version
-        } else
-            arguments[4] = "3"; // odbc version
-        SQLServerResultSet rs;
-        if (connection.isKatmaiOrLater())
-            rs = getResultSetWithProvidedColumnNames(catalog, CallableHandles.SP_COLUMNS, arguments,
-                    getColumnsColumnNamesKatmai);
-        else
-            rs = getResultSetWithProvidedColumnNames(catalog, CallableHandles.SP_COLUMNS, arguments,
-                    getColumnsColumnNames);
-        // Hook in a filter on the DATA_TYPE column of the result set we're
-        // going to return that converts the ODBC values from sp_columns
-        // into JDBC values.
-        rs.getColumn(5).setFilter(new DataTypeFilter());
+            return rs;
+        } else {
+            /**
+             * Can't actually switchCatalogs on Azure DW. This is here to keep consistency in behavior with SQL Azure DB
+             * when user provides a different catalog than the one they're currently connected to. Will throw exception
+             * when it's different and do nothing if it's the same/null.
+             */
+            switchCatalogs(catalog);
```

Comment:
This is else block and only for DW, you can remove `switchCatalogs` call here. Also the comment above.

Reply:
Was asked to add that to maintain consistency with Azure SQL.

Reply:
How about calling `switchCatalogs(catalog);` only if the server `!isAzure()`?

Reply:
You may optionally check for !isAzureDB() in the if block to avoid extra call for Azure DB, but here in else block that we know applies to only DW it makes no sense.

Reply:
I see one reason why this call is useful is to validate if the catalog user specified is correct or not, as this is public API. How about if we verify Catalog name to match database name and throw exception otherwise without the need to contact database for Azure DB/DW/Managed Instance and otherwise let switch catalog? Don't think that would cause much of performance hit since we are already making call to get database type.

Reply:
I see your point. The same table can exist in both databases, so we definitely need to verify the specified catalog name.

Reply:
This change isn't the purpose of this PR and if we do decide to do something like this, it should be applied to every method in this class. As it stands calling `switchCatalogs` just throws an error indicating that catalog switching isn't allowed on Azure, which is basically what we want. We can't check this property without contacting the server because isAzure/DW/MI all contact the server, and this API is meant to work on other databases when on a regular server.

Reply:
Maybe we can leave this change for now then, and handle it in a separate PR. For now, please keep this statement to 'switchCatalog' out of 'if' block as it's called in both cases eitherway.

We may modify the method itself to work differently only for Azure Databases.. to avoid DB call and validate catalog name explicitly, allowing acceptable values only.